_posts/2024-08-27-cheri-myths-safe-languages.markdown

```diff
@@ -0,0 +1,157 @@
+---
+layout: post
+title:  "CHERI Myths: I don't need CHERI if I have safe languages"
+date:   2024-08-27
+categories: cheri myths
+author: David Chisnall
+---
+
+There's a recurring myth that CHERI and safe languages are solving the same problems and that, if you have one, you don't need the other.
+When I joined the CHERI project in 2012, my primary motivation was producing hardware that enabled safe interoperability between languages, so I never felt that safe languages and CHERI were in tension.
+Quite the reverse: My work on CHERI was motivated by a desire to enable safe language adoption.
+
+# Safety is a property of entire programs
+
+The [WannaCry ransomware attack](https://en.wikipedia.org/wiki/WannaCry_ransomware_attack), which cost billions of dollars, was enabled by a single use-after-free bug.
+The [CrowdStrike incident](https://en.wikipedia.org/wiki/CrowdStrike#2024_CrowdStrike_incident), which also cost billions of dollars, was the result of an uninitialised-use bug (compounded by a lot of operational failures).
+
+Memory-safety bugs are particularly bad because they step outside of the language's abstract machine.
+After a program passes a memory-safety bug, it may do anything.
+It may access objects that should not be reachable, modify immutable objects, and even execute instructions out of sequence to do things that are not part of the source-code description of what the program should do at all.
+
+A single memory-safety bug can be all that an attacker needs to get an arbitrary-code execution exploit.
+Rewriting 99% of a program in a safe language but leaving just one memory-safety bug does not prevent at-scale deployment of malware.
+
+This was one of the problems that CHERI aimed to solve, in two ways.
+At the fine granularity, CHERI allows memory-safety properties to be enforced in every language.
+This includes assembly.
+The switcher is the most privileged part of a CHERIoT system and is written in assembly (it has to do things like save all registers, which are not possible in any high-level language).
+In spite of its privilege, it remains bound by the memory-safety rules.
+It cannot access any compartments other than the caller and callee.
+It cannot access any stacks other than the running one and the scheduler's interrupt stack.
```

Comment:
The two examples are indeed memory safety properties, but I might want the connection spelled out in more detail: "Despite that the CHERIoT linker densely packs compartments' global sections, one next to the other, the switcher remains bound by CHERI's spatial rules and has access only to the caller and callee compartment.  Similarly, stacks are again arranged sequentially, but the switcher always has access to only two: the active one (in the register file) and the scheduler's interrupt stack (pointed to by the switcher's internal state)."

Reply:
"And its internal state", right?